packages/ab-testing/src/index.ts

```diff
@@ -25,29 +30,74 @@ function getModuloValue(experiment: string, userId: number | string): number {
     return crc32.calculate(String(userId), crc32.calculate(experiment)) % 100;
 }
 
+function validateAllocationCriteria(filters: ForceInclude, userProfile: UserProfile): boolean {
+    let itemMatchesCriteria = true;
+    for (const key in filters) {
+        if (!filters[key].includes(userProfile[key])) {
+            itemMatchesCriteria = false;
+        }
+    }
+
+    return itemMatchesCriteria;
+}
+
+function validateUserWithinAllocationRange(userSegmentNum: number, range: AllocationV1) {
+    for (const allocation of range) {
+        if (allocation[0] <= userSegmentNum && userSegmentNum < allocation[1]) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+export function validateAllocation(
+    cohort: Cohort,
+    userProfile: UserProfile,
+    configName: string,
+    userId: number | string
+): boolean {
+    const userSegmentNum = getModuloValue(configName, userId);
+    if (cohort.allocation) {
+        if (validateUserWithinAllocationRange(userSegmentNum, cohort.allocation)) {
+            if (cohort.allocation_criteria) {
+                return validateAllocationCriteria(cohort.allocation_criteria, userProfile);
+            }
+            return true;
+        }
+    }
+
+    return false;
+}
+
+function validateForceInclude(cohort: Cohort, userProfile: UserProfile): boolean {
+    if (cohort.force_include) {
+        for (const key in cohort.force_include) {
+            if (cohort.force_include[key].includes(userProfile[key])) {
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
 function matchUserCohort(
     experimentConfig: Experiment,
     userId: number | string,
     userProfile: { [s: string]: string }
 ): string {
-    const userSegmentNum = getModuloValue(experimentConfig.name, userId);
     let allocatedCohort = 'control';
     for (const cohort of experimentConfig.cohorts) {
-        if (cohort.force_include) {
-            for (const key in cohort.force_include) {
-                if (cohort.force_include[key].includes(userProfile[key])) {
-                    return cohort.name;
-                }
-            }
+        if (validateForceInclude(cohort, userProfile)) {
+            return cohort.name;
         }
-        if (allocatedCohort === 'control') {
```

Comment:
this `if` here is also contributes to our perf goal

Reply:
The if here is irrelevant as we are returning cohort.name in force include. If the force include check passes, it will simply break out of the loop and the entire func. Otherwise, it will continue trying to validate allocation, which is why the if statement here is unnecessary.

Reply:
consider this case

```yaml
cohorts:
  - name: control
  - name: group1
     allocation: [0, 100]
     allocation_criteria:
       email_domain:
         - apple.com
  - name: group2
     allocation: [0, 100]
     allocation_criteria:
       email_domain:
         - google.com
  - name: group3
    force_include:
      email:
        - vip1@apple.com
  - name: group4
    force_include:
      email:
        - vip2@apple.com
```

assume the user `vip1@apple.com`, the code would evaluate starting with `group1` which gives a match, and then skip `group2`'s allocation section entirely, **this is the case where `if` optimizes**; then the code continues with `force_include` section in `group3` which yields an early return with `group3`. `group4` is also skipped entirely.

Reply:
can we add a unit test to this case @miguel-gatmaytan? it seems like it's not covered right now

Reply:
Unfortunately 😂  UT will cover these... as these are purely perf optimizations -- the logic remains correct without the optimizations.

Reply:
@ZigZagT I've updated the PR based on your comments, see:

https://github.com/appannie/ab-testing/pull/54/commits/d18130edb0e1a82ff30a3ca6bc0999813c666a21

I moved the userSegmentNum generation back into the ```matchUserCohort```, and I also re-introduced the if statement to check if the allocatedCohort is still control to skip calling the ```validateAllocation``` function if a cohort has already been allocated.

Reply:
> can we add a unit test to this case @miguel-gatmaytan? it seems like it's not covered right now

as @ZigZagT states, the coverage is currently 100% and all paths are covered. This is just a perf issue